### PR TITLE
Added asynchronous plotting to TensorFlow

### PR DESCRIPTION
**Pull request by jonashen**
Unfortunately, async plotting for TensorFlow requires the use of threading.Thread instead of multiprocessing.Process. Thus, on Linux machines, there will be a slight delay when you run the algorithm that seems like nothing is happening, but once the window loads everything works perfectly and asynchronously.

See: #55

### CODE REVIEW

File: rllab/misc/tabulate.py [hunk 1]
```diff
@@ -1,8 +1,9 @@
 """Pretty-print tabular data."""
 
 from collections import namedtuple
+import functools
```
Comment (gonzaiva) [inline rllab/misc/tabulate.py]: It may be faster to only import reduce like `from itertools import reduce`

**Comment by ryanjulian** [comment]
You may want to wait on #59

File: sandbox/rocky/tf/algos/batch_polopt.py [hunk 2]
```diff
@@ -165,3 +167,7 @@ def get_itr_snapshot(self, itr, samples_data):
 
     def optimize_policy(self, itr, samples_data):
         raise NotImplementedError
+
+    def update_plot(self):
```
Comment (gonzaiva) [inline sandbox/rocky/tf/algos/batch_polopt.py]: Not used. Remove.

File: sandbox/rocky/tf/plotter/plotter.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+__all__ = ["Plotter"]
+
+
+class Op(Enum):
+    STOP = 0
+    UPDATE = 1
+    DEMO = 2
+
+
+Message = namedtuple("Message", ["op", "args", "info"])
```
Comment (ryanjulian) [inline sandbox/rocky/tf/plotter/plotter.py]: info? the python convention would be kwargs (ignore what gym does)

File: sandbox/rocky/tf/plotter/plotter.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+import platform
+from queue import Empty, Queue
+from threading import Thread
+
+import numpy as np
+import tensorflow as tf
+
+from rllab.sampler.utils import rollout
+
+__all__ = ["Plotter"]
```
Comment (gonzaiva) [inline sandbox/rocky/tf/plotter/plotter.py]: This is not required if  * is not used in the import. See [here](https://docs.python.org/3.5/tutorial/modules.html#importing-from-a-package) for further information.

File: sandbox/rocky/tf/plotter/plotter.py (hunk 3 above)
Comment (ryanjulian) [reply sandbox/rocky/tf/plotter/plotter.py]: To add more color, the intention of the interface is this.

All python function definitions can be written as
```python
def foo(*args, **kwargs):
    # do some stuff
```
where `args` is an ordered list of unnamed arguments and `kwargs` is a dictionary of keyword arguments. You can even call any function like this, even if it wasn't declared this way.
```
def foo(bar, baz=0):
    return bar + baz

args = [1]
kwargs = {"baz": 1}
print(foo(*args, **kwargs)) # prints 2
```

Now imagine that this interface gets complicated, and we split the implementation of DEMO, UPDATE, and STOP into functions.
If `Message` has op, args, and kwargs, we can implement dispatch for *any function*:
```python
if Op.FOO in msgs:
    msg = msgs[Op.FOO]
    do_foo(*msg.args, **msg.kwargs)
```

**Comment by ryanjulian** [comment]
I apologize for asking you to fix the imports for just this PR--I didn't realize how many files you'd have to touch.

File: rllab/tf/algos/npo.py [hunk 5]
```diff
@@ -28,7 +28,7 @@ def __init__(self,
         self.name = name
         super(NPO, self).__init__(**kwargs)
 
-    @overrides
+    # @overrides
```
Comment (ryanjulian) [inline rllab/tf/algos/npo.py]: uhh why is this commented out?

File: rllab/tf/algos/npo.py [hunk 6]
```diff
@@ -105,7 +105,7 @@ def init_opt(self):
                 constraint_name="mean_kl")
             return dict()
 
-    @overrides
+    # @overrides
```
Comment (ryanjulian) [inline rllab/tf/algos/npo.py]: why commented out?

**Pull request merged**